src/BackgroundTask/BackgroundTaskDeactivator.php

```diff
@@ -0,0 +1,171 @@
+	private $events_to_deactivate = [];
+
+	/**
+	 * Name of the plugin as WordPress is expecting it.
+	 *
+	 * This should usually have the form "amp/amp.php".
+	 *
+	 * @var string
+	 */
+	private $plugin_file;
+
+	/**
+	 * Check whether the conditional object is currently needed.
+	 *
+	 * @return bool Whether the conditional object is needed.
+	 */
+	public static function is_needed() {
+		return is_admin() || wp_doing_cron();
+	}
+
+	/**
+	 * Class constructor.
+	 */
+	public function __construct() {
+		$this->plugin_file = plugin_basename( AMP__DIR__ . '/amp.php' );
+	}
+
+	/**
+	 * Register the service with the system.
+	 *
+	 * @return void
+	 */
+	public function register() {
+		add_action( "network_admin_plugin_action_links_{$this->plugin_file}", [ $this, 'add_warning_sign_to_network_deactivate_action' ], 10, 1 );
+		add_action( 'plugin_row_meta', [ $this, 'add_warning_to_plugin_meta' ], 10, 2 );
+	}
+
+	/**
+	 * Get warning icon markup.
+	 *
+	 * @return string Warning icon markup.
+	 */
+	private function get_warning_icon() {
+		return sprintf( '<span style="vertical-align: middle">%s</span>', Icon::warning()->to_html() );
+	}
+
+	/**
+	 * Adds an event to the deactivate queue.
+	 *
+	 * @param string $event_name The event name.
+	 */
+	public function add_event( $event_name ) {
+		if ( ! in_array( $event_name, $this->events_to_deactivate, true ) ) {
+			$this->events_to_deactivate[] = $event_name;
+		}
+	}
+
+	/**
+	 * Run deactivation logic.
+	 *
+	 * This should be hooked up to the WordPress deactivation hook.
+	 *
+	 * @param bool $network_wide Whether the deactivation was done network-wide.
+	 * @return void
+	 */
+	public function deactivate( $network_wide ) {
+		if ( $network_wide && is_multisite() && ! wp_is_large_network( 'sites' ) ) {
+			foreach ( get_sites(
+				[
+					'fields' => 'ids',
+					'number' => 0,
+				]
+			) as $blog_id ) {
+				switch_to_blog( $blog_id );
+
+				foreach ( $this->events_to_deactivate as $event_name ) {
+					wp_clear_scheduled_hook( $event_name );
+				}
+
+				restore_current_blog();
+			}
+		} else {
+			foreach ( $this->events_to_deactivate as $event_name ) {
+				wp_clear_scheduled_hook( $event_name );
+			}
+		}
+	}
+
+	/**
+	 * Add a warning sign to the network deactivate action on the network plugins screen.
+	 *
+	 * @param string[] $actions     An array of plugin action links. By default this can include 'activate',
+	 *                              'deactivate', and 'delete'.
+	 * @return string[]
+	 */
+	public function add_warning_sign_to_network_deactivate_action( $actions ) {
+		if ( ! wp_is_large_network() ) {
+			return $actions;
+		}
+
+		if ( ! array_key_exists( 'deactivate', $actions ) ) {
+			return $actions;
+		}
+
+		wp_enqueue_style( 'amp-icons' );
+
+		$warning_icon = $this->get_warning_icon();
+		if ( false === strpos( $actions['deactivate'], $warning_icon ) ) {
+			$actions['deactivate'] = preg_replace( '#(?=</a>)#i', ' ' . $warning_icon, $actions['deactivate'] );
+		}
```

Comment:
@westonruter I noticed you were the original author of this snippet of code.

How about instead of only appending the warning icon to the anchor text, we use a confirmation dialog to verify they want to deactivate? For example:

```suggestion
		if ( preg_match( '/<a\s+.*?id\s*=\s*"([^"]+)"/', $actions['deactivate'], $matches ) ) {
			$anchor_id = $matches[1];
			$confirm_text = __( 'Deactivation will leave orphaned scheduled events behind. Do you want to continue?', 'amp' );

			$script = "
				document.getElementById( '{$anchor_id}' ).addEventListener( 'click', function ( event ) {
					if ( ! confirm( '{$confirm_text}' ) ) {
						event.preventDefault();
					}
				} )
			";

			wp_add_inline_script( 'wp-dom-ready', $script );
		}
```

And here's it in action:

![](https://user-images.githubusercontent.com/16200219/100193984-ceb41500-2eec-11eb-8699-88423e4b8e22.gif)

Reply:
Actually, the warning was originally added by @schlessera in eaf8259b87cc76a0782599ebea5583d8e5d143d2, although it was further iterated on.

Adding a `confirm()` would work. The only downside would be if you select multiple plugins and do a bulk deactivation: you'd easily miss that the plugin has special considerations for deactivation.

Reply:
I have to say I'm pretty agnostic on this. We could do both (the icon and the confirm), make no change, or make a new issue for further discussion. Which way are you leaning @westonruter @pierlon ?

Reply:
> The only downside would be if you select multiple plugins and do a bulk deactivation: you'd easily miss that the plugin has special considerations for deactivation.

@westonruter The `confirm()` could also be applied to that also, no?

I agree with @johnwatkins0 with regards to moving this discussion to its own issue.